src/v/iceberg/tests/manifest_serialization_test.cc

```diff
@@ -0,0 +1,253 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "base/vlog.h"
+#include "bytes/iobuf.h"
+#include "iceberg/manifest_entry.h"
+#include "iceberg/manifest_file.h"
+
+#include <seastar/core/temporary_buffer.hh>
+
+#include <avro/Stream.hh>
+#include <gtest/gtest.h>
+
+using namespace iceberg;
+
+namespace {
+
+ss::logger icelog("icebergtest");
+
+// Near-identical implementation of avro::MemoryOutputStream, but backed by an
+// iobuf that can be released.
+class avro_iobuf_ostream : public avro::OutputStream {
+public:
+    explicit avro_iobuf_ostream(size_t chunk_size)
+      : chunk_size_(chunk_size)
+      , available_(0)
+      , byte_count_(0) {}
+    ~avro_iobuf_ostream() override = default;
+
+    iobuf&& release() {
+        buf_.trim_back(buf_.size_bytes() - byte_count_);
+        available_ = 0;
+        byte_count_ = 0;
+        vlog(icelog.trace, "ostream::release({})", buf_.size_bytes());
+        return std::move(buf_);
+    }
+
+    // If there's no available space in the buffer, allocates `chunk_size_`
+    // bytes at the end of the buffer.
+    //
+    // Returns the current position in the buffer, and the available remaining
+    // space.
+    bool next(uint8_t** data, size_t* len) final {
+        vlog(icelog.trace, "ostream::next()");
+        if (available_ == 0) {
+            buf_.append(ss::temporary_buffer<char>{chunk_size_});
+            available_ = chunk_size_;
+        }
+        auto back_frag = buf_.rbegin();
+        *data = reinterpret_cast<uint8_t*>(
+          back_frag->share(chunk_size_ - available_, available_).get_write());
+        *len = available_;
+        byte_count_ += available_;
+        available_ = 0;
+        return true;
+    }
+
+    void backup(size_t len) final {
```

Comment:
Do we need to ensure that `len <= chunksize - available`?

Reply:
It's probably a good practice, but for now I don't think so. This impl is pretty much copying the chunked in-memory streams which don't do this check[1], and looking at the call site, it appears this will only be called with a value under the chunk size[2] (note, the len returned from next() is under the chunk size).

[1] https://github.com/apache/avro/blob/8dcd9c7ac253ba3db15efaa2eae26b2617206362/lang/c%2B%2B/impl/Stream.cc#L26
[2] https://github.com/apache/avro/blob/8dcd9c7ac253ba3db15efaa2eae26b2617206362/lang/c%2B%2B/impl/DataFile.cc#L373